src/lib/core/clock_lowres.c

```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright 2010-2022, Tarantool AUTHORS: please see AUTHORS file.
+ *
+ * Redistribution and use in source and binary forms, with or
+ * without modification, are permitted provided that the following
+ * conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above
+ *    copyright notice, this list of conditions and the
+ *    following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above
+ *    copyright notice, this list of conditions and the following
+ *    disclaimer in the documentation and/or other materials
+ *    provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY <COPYRIGHT HOLDER> ``AS IS'' AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+ * TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL
+ * <COPYRIGHT HOLDER> OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
+ * INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR
+ * BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+ * LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
+ * THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ */
+#include <signal.h>
+#include <string.h>
+#include <say.h>
+#include <math.h>
+#include "clock.h"
+
+/**
+ * Granularity in seconds.
+ * SIGALRM handler interrupts some syscalls (open, read, etc.), so
+ * too small granularity is not recommended.
+ */
+#define LOW_RES_GRANULARITY 0.05
+
+double low_res_clock = 0.0;
+
+/** A tick of low_res_clock, SIGALRM handler. */
+static void
+clock_lowres_tick(int signum)
+{
+	(void)signum;
+	low_res_clock += LOW_RES_GRANULARITY;
+}
+
+void
+clock_lowres_signal_init(void)
+{
+	struct sigaction sa;
+	memset(&sa, 0, sizeof(sa));
+	sa.sa_handler = clock_lowres_tick;
+	sa.sa_flags = SA_RESTART;
+	if (sigaction(SIGALRM, &sa, NULL) == -1)
+		panic_syserror("sigaction");
+
+	struct itimerval timer;
+	long usec = floor(LOW_RES_GRANULARITY * 1e6);
+	long sec = floor(LOW_RES_GRANULARITY);
+	timer.it_interval.tv_usec = usec;
+	timer.it_interval.tv_sec = sec;
+	timer.it_value.tv_usec = usec;
+	timer.it_value.tv_sec = sec;
+	if (setitimer(ITIMER_REAL, &timer, NULL) == -1)
+		panic_syserror("setitimer");
+}
+
+void
+clock_lowres_signal_reset(void)
+{
+	struct itimerval timer;
+	memset(&timer, 0, sizeof(timer));
+	if (setitimer(ITIMER_REAL, &timer, NULL) == -1)
+		say_syserror("clock_lowres: reset setitimer");
+
+	struct sigaction sa;
+	memset(&sa, 0, sizeof(sa));
+	sa.sa_handler = SIG_DFL;
+	if (sigaction(SIGALRM, &sa, 0) == -1)
```

Comment:
Perhaps it would be better to save the old `sigaction` to a static variable in `clock_lowres_signal_init` and restore it here?

RFC @locker

Reply:
done